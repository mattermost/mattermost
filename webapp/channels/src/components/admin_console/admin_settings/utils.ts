// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {AdminConfig, EnvironmentConfig} from '@mattermost/types/config';

function getConfigValue(config: AdminConfig | Partial<EnvironmentConfig>, path: string) {
    const pathParts = path.split('.');

    return pathParts.reduce((obj: object | null, pathPart) => {
        if (!obj) {
            return null;
        }
        return obj[(pathPart as keyof object)];
    }, config);
}

export function isSetByEnv(environmentConfig: Partial<EnvironmentConfig>, path: string) {
    return Boolean(getConfigValue(environmentConfig, path));
}

export const parseIntNonNegative = (str: string | number, defaultValue?: number) => {
    const n = typeof str === 'string' ? parseInt(str, 10) : str;

    if (isNaN(n) || n < 0) {
        if (defaultValue) {
            return defaultValue;
        }
        return 0;
    }

    return n;
};

export const parseIntZeroOrMin = (str: string | number, minimumValue = 1) => {
    const n = typeof str === 'string' ? parseInt(str, 10) : str;

    if (isNaN(n) || n < 0) {
        return 0;
    }
    if (n > 0 && n < minimumValue) {
        return minimumValue;
    }

    return n;
};

export const parseIntNonZero = (str: string | number, defaultValue?: number, minimumValue = 1) => {
    const n = typeof str === 'string' ? parseInt(str, 10) : str;

    if (isNaN(n) || n < minimumValue) {
        if (defaultValue) {
            return defaultValue;
        }
        return 1;
    }

    return n;
};
