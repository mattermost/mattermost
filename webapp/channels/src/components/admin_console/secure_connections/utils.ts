// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {LocationDescriptor} from 'history';
import {DateTime, Interval} from 'luxon';
import {useCallback, useEffect, useMemo, useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';

import type {ClientError} from '@mattermost/client';
import type {Channel} from '@mattermost/types/channels';
import {isRemoteClusterPatch, type RemoteCluster, type RemoteClusterPatch} from '@mattermost/types/remote_clusters';
import type {SharedChannelRemote} from '@mattermost/types/shared_channels';
import type {Team} from '@mattermost/types/teams';
import type {IDMappedObjects, RelationOneToOne} from '@mattermost/types/utilities';

import {ChannelTypes} from 'mattermost-redux/action_types';
import {getChannel as fetchChannel} from 'mattermost-redux/actions/channels';
import {Client4} from 'mattermost-redux/client';
import {getChannel} from 'mattermost-redux/selectors/entities/channels';
import {getActiveTeamsList, getTeam} from 'mattermost-redux/selectors/entities/teams';

import type {ActionFuncAsync} from 'types/store';

export const useRemoteClusters = () => {
    const [remoteClusters, setRemoteClusters] = useState<RemoteCluster[]>();
    const [loadingState, setLoadingState] = useState<boolean | ClientError>(true);
    const {loading, error} = loadingStatus(loadingState);

    const fetch = async () => {
        setLoadingState(true);
        try {
            const data = await Client4.getRemoteClusters({excludePlugins: true});
            setRemoteClusters(data?.length ? data : undefined);
            setLoadingState(false);
        } catch (err) {
            setLoadingState(err);
        }
    };

    useEffect(() => {
        fetch();
    }, []);

    return [remoteClusters, {loading, fetch, error}] as const;
};

export const useRemoteClusterEdit = (remoteId: string | 'create', initRemoteCluster: RemoteCluster | undefined) => {
    const editing = remoteId !== 'create';

    const [currentRemoteCluster, setCurrentRemoteCluster] = useState<RemoteCluster | undefined>(initRemoteCluster);
    const [patch, setPatch] = useState<Partial<RemoteClusterPatch>>({});
    const [loading, setLoading] = useState<TLoadingState>(editing && !currentRemoteCluster);
    const [saving, setSaving] = useState<TLoadingState>(false);

    const hasChanges = Object.keys(patch).length > 0;

    useEffect(() => {
        if (!editing) {
            return;
        }
        (async () => {
            try {
                const data = await Client4.getRemoteCluster(remoteId);
                setCurrentRemoteCluster(data);
                setLoading(false);
                setPatch({});
            } catch (err) {
                setSaving(err);
            }
        })();
    }, [remoteId]);

    const applyPatch = (patch: Partial<RemoteClusterPatch>) => {
        setPatch((current) => ({...current, ...patch}));
    };

    const save = async () => {
        if (currentRemoteCluster && isRemoteClusterPatch(patch)) {
            setSaving(true);
            try {
                const data = await Client4.patchRemoteCluster(remoteId, patch);
                setCurrentRemoteCluster(data);
                setSaving(false);
                setPatch({});
            } catch (err) {
                setSaving(err);
            }
            setSaving(false);
        }
    };

    return [
        {...currentRemoteCluster, ...patch},
        {
            applyPatch,
            save,
            hasChanges,
            loading,
            saving,
            currentRemoteCluster,
            patch,
        },
    ] as const;
};

export const useSharedChannelRemotes = (remoteId: string) => {
    const [remotes, setRemotes] = useState<RelationOneToOne<Channel, SharedChannelRemote>>();
    const [loadingState, setLoadingState] = useState<TLoadingState>(true);

    const loading = isPendingState(loadingState);
    const error = !loading && loadingState;

    const fetch = async () => {
        setLoadingState(true);
        try {
            const data = await Client4.getSharedChannelRemotes(remoteId, {include_deleted: true, include_unconfirmed: true});

            setRemotes(data?.reduce<typeof remotes>((state, remote) => {
                state![remote.channel_id] = remote;

                return state;
            }, {}));
            setLoadingState(false);
        } catch (error) {
            setLoadingState(error);
        }
    };

    useEffect(() => {
        fetch();
    }, [remoteId]);

    return [remotes, {loading, error, fetch}] as const;
};

const remoteRow = (remote: SharedChannelRemote, channel: Channel, team?: Team) => {
    return {...remote, display_name: channel.display_name, team_display_name: team?.display_name ?? ''};
};

export type SharedChannelRemoteRow = SharedChannelRemote & Pick<Channel, 'display_name'> & {team_display_name: string};
export const useSharedChannelRemoteRows = (remoteId: string, opts: {filter: 'home' | 'remote' | undefined}) => {
    const [sharedChannelRemotes, setSharedChannelRemotes] = useState<SharedChannelRemoteRow[]>();
    const [loadingState, setLoadingState] = useState<TLoadingState>(true);
    const dispatch = useDispatch();

    const loading = isPendingState(loadingState);
    const error = !loading && loadingState;

    const fetch = useCallback(async () => {
        if (opts.filter === undefined) {
            // wait for a filter
            return;
        }

        setLoadingState(true);
        dispatch<ActionFuncAsync<IDMappedObjects<SharedChannelRemoteRow>>>(async (dispatch, getState) => {
            const collected: IDMappedObjects<SharedChannelRemoteRow> = {};
            const missing: SharedChannelRemote[] = [];

            try {
                const data = await Client4.getSharedChannelRemotes(remoteId, {include_unconfirmed: true, exclude_remote: opts.filter === 'home', exclude_home: opts.filter === 'remote'});
                let state = getState();
                let getMyChannelsOnce: undefined | ((remote: SharedChannelRemote) => Promise<Channel | undefined>) = async (firstNotFound: SharedChannelRemote) => {
                    const channels = await Client4.getAllTeamsChannels();
                    dispatch({
                        type: ChannelTypes.RECEIVED_ALL_CHANNELS,
                        data: channels,
                    });
                    state = getState();

                    getMyChannelsOnce = undefined; // once-only

                    // return triggering not-found remote
                    return getChannel(state, firstNotFound.channel_id);
                };

                // first-past, get known channels or do initial load on first not-found channel
                for (const remote of data) {
                    // eslint-disable-next-line no-await-in-loop
                    const channel = getChannel(state, remote.channel_id) ?? await getMyChannelsOnce?.(remote);

                    if (!channel) {
                        // collect all remotes with missing channels
                        missing?.push(remote);
                        continue;
                    }

                    const team = getTeam(state, channel.team_id);
                    collected[remote.id] = remoteRow(remote, channel, team);
                }

                // fetch missing channels individually
                if (missing.length) {
                    // TODO: performance; consider adding sharedchannelremotes search param to api/v4/channels
                    await Promise.allSettled(missing.map((remote) => dispatch(fetchChannel(remote.channel_id))));
                    state = getState();

                    for (const remote of missing) {
                        const channel = getChannel(state, remote.channel_id);

                        if (!channel) {
                            continue;
                        }

                        const team = getTeam(state, channel.team_id);
                        collected[remote.id] = remoteRow(remote, channel, team);
                    }
                }

                const rows = Object.values(collected);

                setSharedChannelRemotes(rows.length ? rows : undefined);
                setLoadingState(false);
            } catch (error) {
                setLoadingState(error);
                return {error};
            }

            return {data: collected};
        });
    }, [remoteId, opts.filter]);

    useEffect(() => {
        fetch();
    }, [remoteId, opts.filter]);

    return [sharedChannelRemotes, {loading, error, fetch}] as const;
};

export const useTeamOptions = () => {
    const teams = useSelector(getActiveTeamsList);
    const teamsById = useMemo(() => teams.reduce<IDMappedObjects<Team>>((teams, team) => ({...teams, [team.id]: team}), {}), [teams]);
    return teamsById;
};

export const getEditLocation = (rc: RemoteCluster): LocationDescriptor<RemoteCluster> => {
    return {pathname: `/admin_console/site_config/secure_connections/${rc.remote_id}`, state: rc};
};

export const getCreateLocation = (): LocationDescriptor<RemoteCluster> => {
    return {pathname: '/admin_console/site_config/secure_connections/create'};
};

const SiteURLPendingPrefix = 'pending_';
export const isConfirmed = (rc: RemoteCluster) => Boolean(rc.site_url && !rc.site_url.startsWith(SiteURLPendingPrefix));
export const isConnected = (rc: RemoteCluster) => Interval.before(DateTime.now(), {minutes: 5}).contains(DateTime.fromMillis(rc.last_ping_at));

export type TLoadingState<TError extends Error = ClientError> = boolean | TError;
export const isPendingState = <T extends Error>(loadingState: TLoadingState<T>) => loadingState === true;
export const isErrorState = <T extends Error>(loadingState: TLoadingState<T>): loadingState is T => loadingState instanceof Error;

const loadingStatus = <T extends Error>(loadingState: TLoadingState<T>) => {
    const loading = isPendingState(loadingState);
    const error = isErrorState(loadingState) ? loadingState : undefined;

    return {error, loading};
};
