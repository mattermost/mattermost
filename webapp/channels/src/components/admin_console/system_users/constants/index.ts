// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export enum ColumnNames {
    username = 'usernameColumn',
    email = 'emailColumn',
    createAt = 'createAtColumn',
    lastLoginAt = 'lastLoginColumn',
    lastStatusAt = 'lastStatusAtColumn',
    lastPostDate = 'lastPostDateColumn',
    daysActive = 'daysActiveColumn',
    totalPosts = 'totalPostsColumn',
    actions = 'actionsColumn',
}

export enum StatusFilter {
    Any = 'any',
    Active = 'active',
    Deactivated = 'deactivated',
}

export enum RoleFilters {
    Any = 'any',
    Admin = 'system_admin',
    Member = 'system_user',
    Guest = 'system_guest',
}

export enum TeamFilters {
    AllTeams = 'teams_filter_for_all_teams',
    NoTeams = 'teams_filter_for_no_teams',
}
