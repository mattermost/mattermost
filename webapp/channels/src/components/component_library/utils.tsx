// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

function buildPropsLists(inputPossibilities: {[x: string]: any[]}): Array<{[x: string]: any}> {
    const keys = Object.keys(inputPossibilities);
    if (!keys.length) {
        return [{}];
    }

    const selectedKey = keys[0];
    const restPossibilities = {...inputPossibilities};
    delete restPossibilities[selectedKey];
    const subProps = buildPropsLists(restPossibilities);
    const result: Array<{[x: string]: any}> = [];
    inputPossibilities[selectedKey].forEach((v) => {
        subProps.forEach((rest) => {
            result.push({...rest, [selectedKey]: v});
        });
    });

    return result;
}

function buildPropString(inputProps: {[x: string]: any}) {
    const propKeys = Object.keys(inputProps);
    if (!propKeys.length) {
        return undefined;
    }

    const result = [(<>{'PROPS: '}</>)];
    propKeys.forEach((v) => {
        result.push((<><b>{v}</b>{`: ${inputProps[v]}, `}</>));
    });
    return result;
}

export function buildComponent(
    Component: React.ComponentType<any>,
    propPossibilities: {[x: string]: any[]},
    dropdownPossibilities: Array<{[x: string]: string[]} | undefined>,
    setProps: Array<{[x: string]: any} | undefined>,
) {
    const res: React.ReactNode[] = [];
    let currentPropPossibilities = {...propPossibilities};
    dropdownPossibilities.forEach((v) => {
        if (v) {
            currentPropPossibilities = {
                ...currentPropPossibilities,
                ...v,
            };
        }
    });

    const propsVariations = buildPropsLists(currentPropPossibilities);
    let builtSetProps = {};
    setProps.forEach((v) => {
        if (v) {
            builtSetProps = {
                ...builtSetProps,
                ...v,
            };
        }
    });
    propsVariations.forEach((v) => {
        const propString = buildPropString(v);
        res.push(
            <>
                {Boolean(propString) && <p>{propString}</p>}
                <Component
                    {...builtSetProps}
                    {...v}
                />
            </>,
        );
    });
    return res;
}
