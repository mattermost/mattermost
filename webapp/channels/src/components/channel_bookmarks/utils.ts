// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {useEffect, useMemo, useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';

import type {Channel} from '@mattermost/types/channels';
import type {GlobalState} from '@mattermost/types/store';

import {Permissions} from 'mattermost-redux/constants';
import {getChannelBookmarks} from 'mattermost-redux/selectors/entities/channel_bookmarks';
import {getChannel, getMyChannelMember} from 'mattermost-redux/selectors/entities/channels';
import {getConfig, getFeatureFlagValue, getLicense} from 'mattermost-redux/selectors/entities/general';
import {haveIChannelPermission} from 'mattermost-redux/selectors/entities/roles';
import {insertWithoutDuplicates} from 'mattermost-redux/utils/array_utils';

import {fetchChannelBookmarks, reorderBookmark} from 'actions/channel_bookmarks';
import {loadCustomEmojisIfNeeded} from 'actions/emoji_actions';

import Constants from 'utils/constants';
import {trimmedEmojiName} from 'utils/emoji_utils';
import {canUploadFiles, isPublicLinksEnabled} from 'utils/file_utils';

export const MAX_BOOKMARKS_PER_CHANNEL = 50;

const {OPEN_CHANNEL, PRIVATE_CHANNEL, GM_CHANNEL, DM_CHANNEL} = Constants as {OPEN_CHANNEL: 'O'; PRIVATE_CHANNEL: 'P'; GM_CHANNEL: 'G'; DM_CHANNEL: 'D'};

type TAction = 'add' | 'edit' | 'delete' | 'order';
type TActionKey = `${TAction}${typeof OPEN_CHANNEL | typeof PRIVATE_CHANNEL}`;

const key = (a: TAction, c: typeof OPEN_CHANNEL | typeof PRIVATE_CHANNEL): TActionKey => {
    return `${a}${c}`;
};

const BOOKMARK_PERMISSION = {

    // open channel
    [key('add', OPEN_CHANNEL)]: Permissions.ADD_BOOKMARK_PUBLIC_CHANNEL,
    [key('edit', OPEN_CHANNEL)]: Permissions.EDIT_BOOKMARK_PUBLIC_CHANNEL,
    [key('delete', OPEN_CHANNEL)]: Permissions.DELETE_BOOKMARK_PUBLIC_CHANNEL,
    [key('order', OPEN_CHANNEL)]: Permissions.ORDER_BOOKMARK_PUBLIC_CHANNEL,

    // private channel
    [key('add', PRIVATE_CHANNEL)]: Permissions.ADD_BOOKMARK_PRIVATE_CHANNEL,
    [key('edit', PRIVATE_CHANNEL)]: Permissions.EDIT_BOOKMARK_PRIVATE_CHANNEL,
    [key('delete', PRIVATE_CHANNEL)]: Permissions.DELETE_BOOKMARK_PRIVATE_CHANNEL,
    [key('order', PRIVATE_CHANNEL)]: Permissions.ORDER_BOOKMARK_PRIVATE_CHANNEL,
} as const;

export const useChannelBookmarkPermission = (channelId: string, action: TAction) => {
    return useSelector((state: GlobalState) => getHaveIChannelBookmarkPermission(state, channelId, action));
};

export const getHaveIChannelBookmarkPermission = (state: GlobalState, channelId: string, action: TAction) => {
    const channel: Channel | undefined = getChannel(state, channelId);

    if (!channel) {
        return false;
    }

    if (channel.delete_at !== 0) {
        return false;
    }

    const {type} = channel;

    if (type === 'threads') {
        return false;
    }

    if (type === GM_CHANNEL || type === DM_CHANNEL) {
        const myMembership = getMyChannelMember(state, channelId);
        return myMembership?.channel_id === channelId;
    }

    const permission = BOOKMARK_PERMISSION[key(action, type)];

    return channel && permission && haveIChannelPermission(state, channel.team_id, channelId, permission);
};

export const useCanUploadFiles = () => {
    return useSelector((state: GlobalState) => canUploadFiles(getConfig(state)));
};

export const useCanGetPublicLink = () => {
    return useSelector((state: GlobalState) => isPublicLinksEnabled(getConfig(state)));
};

export const useCanGetLinkPreviews = () => {
    return useSelector((state: GlobalState) => getConfig(state).EnableLinkPreviews === 'true');
};

export const getIsChannelBookmarksEnabled = (state: GlobalState) => {
    const isEnabled = getFeatureFlagValue(state, 'ChannelBookmarks') === 'true';

    if (!isEnabled) {
        return false;
    }

    const license = getLicense(state);

    return license?.IsLicensed === 'true';
};

export const useChannelBookmarks = (channelId: string) => {
    const dispatch = useDispatch();
    const bookmarks = useSelector((state: GlobalState) => getChannelBookmarks(state, channelId));

    const order = useMemo(() => {
        return Object.keys(bookmarks).sort((a, b) => bookmarks[a].sort_order - bookmarks[b].sort_order);
    }, [bookmarks]);
    const [tempOrder, setTempOrder] = useState<typeof order>();

    useEffect(() => {
        if (tempOrder) {
            setTempOrder(undefined);
        }
    }, [order]);

    useEffect(() => {
        if (channelId) {
            dispatch(fetchChannelBookmarks(channelId));
        }
    }, [channelId]);

    useEffect(() => {
        const emojis = Object.values(bookmarks).reduce<string[]>((result, {emoji}) => {
            if (emoji) {
                result.push(trimmedEmojiName(emoji));
            }

            return result;
        }, []);

        if (emojis.length) {
            dispatch(loadCustomEmojisIfNeeded(emojis));
        }
    }, [bookmarks]);

    const reorder = async (id: string, prevOrder: number, nextOrder: number) => {
        setTempOrder(insertWithoutDuplicates(order, id, nextOrder));
        const {error} = await dispatch(reorderBookmark(channelId, id, nextOrder));

        if (error) {
            setTempOrder(undefined);
        }
    };

    return {
        bookmarks,
        order: tempOrder ?? order,
        reorder,
    } as const;
};

