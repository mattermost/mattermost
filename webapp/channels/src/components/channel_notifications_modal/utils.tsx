// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {defineMessage, defineMessages, FormattedMessage, type IntlShape} from 'react-intl';

import type {FieldsetCheckbox} from 'components/widgets/modals/components/checkbox_setting_item';
import type {FieldsetRadio} from 'components/widgets/modals/components/radio_setting_item';
import type {FieldsetReactSelect} from 'components/widgets/modals/components/react_select_item';

import {NotificationLevels} from 'utils/constants';
import {optionsOfMessageNotificationSoundsSelect} from 'utils/notification_sounds';

const MuteChannelInputFieldData: FieldsetCheckbox = {
    name: 'mute channel',
    dataTestId: 'muteChannel',
};

const DesktopReplyThreadsInputFieldData: FieldsetCheckbox = {
    name: 'desktop reply threads',
    dataTestId: 'desktopReplyThreads',
};

const MobileReplyThreadsInputFieldData: FieldsetCheckbox = {
    name: 'mobile reply threads',
    dataTestId: 'mobileReplyThreads',
};

export const sameMobileSettingsDesktopInputFieldData: FieldsetCheckbox = {
    name: 'same mobile settings as Desktop',
    dataTestId: 'sameMobileSettingsDesktop',
};

export const IgnoreMentionsInputFieldData: FieldsetCheckbox = {
    name: 'ignore mentions',
    dataTestId: 'ignoreMentions',
};

export const AutoFollowThreadsInputFieldData: FieldsetCheckbox = {
    name: 'auto follow threads',
    dataTestId: 'autoFollowThreads',
};

const defaultMessage = defineMessage({
    id: 'channel_notifications.default',
    defaultMessage: '(default)',
});

export const desktopNotificationInputFieldData = (defaultOption: string, formatMessage: IntlShape['formatMessage']): FieldsetRadio => {
    const messages = defineMessages({
        allMessages: {
            id: 'channelNotifications.desktopNotification.allMessages',
            defaultMessage: 'All new messages {optionalDefault}',

        },
        mention: {
            id: 'channelNotifications.desktopNotification.mention',
            defaultMessage: 'Mentions, direct messages, and keywords only {optionalDefault}',
        },
        nothing: {
            id: 'channelNotifications.desktopNotification.nothing',
            defaultMessage: 'Nothing {optionalDefault}',
        },
    });

    return {
        options: [
            {
                dataTestId: `desktopNotification-${NotificationLevels.ALL}`,
                title: (
                    <FormattedMessage
                        {...messages.allMessages}
                        values={{
                            optionalDefault: defaultOption === NotificationLevels.ALL ? (
                                <FormattedMessage
                                    {...defaultMessage}
                                />) : undefined,
                        }}
                    />
                ),
                name: formatMessage(messages.allMessages, {
                    optionalDefault: defaultOption === NotificationLevels.ALL ? (
                        formatMessage(defaultMessage)
                    ) : undefined,
                }),
                key: `desktopNotification-${NotificationLevels.ALL}`,
                value: NotificationLevels.ALL,
            },
            {
                dataTestId: `desktopNotification-${NotificationLevels.MENTION}`,
                title: (
                    <FormattedMessage
                        {...messages.mention}
                        values={{
                            optionalDefault: defaultOption === NotificationLevels.MENTION ? (
                                <FormattedMessage
                                    {...defaultMessage}
                                />) : undefined,
                        }}
                    />
                ),
                name: formatMessage(messages.mention, {
                    optionalDefault: defaultOption === NotificationLevels.MENTION ? (
                        formatMessage(defaultMessage)
                    ) : undefined,
                }),
                key: `desktopNotification-${NotificationLevels.MENTION}`,
                value: NotificationLevels.MENTION,
            },
            {
                dataTestId: `desktopNotification-${NotificationLevels.NONE}`,
                title: (
                    <FormattedMessage
                        {...messages.nothing}
                        values={{
                            optionalDefault: defaultOption === NotificationLevels.NONE ? (
                                <FormattedMessage
                                    {...defaultMessage}
                                />) : undefined,
                        }}
                    />
                ),
                name: formatMessage(messages.nothing, {
                    optionalDefault: defaultOption === NotificationLevels.NONE ? (
                        formatMessage(defaultMessage)
                    ) : undefined,
                }),
                key: `desktopNotification-${NotificationLevels.NONE}`,
                value: NotificationLevels.NONE,
            },
        ],
    };
};

export const desktopNotificationSoundsCheckboxFieldData: FieldsetCheckbox = {
    name: 'desktopNotificationSoundsCheckbox',
    dataTestId: 'desktopNotificationSoundsCheckbox',
};

export const desktopNotificationSoundsSelectFieldData: FieldsetReactSelect = {
    id: 'desktopNotificationSoundsSelect',
    inputId: 'desktopNotificationSoundsSelectInputId',
    options: optionsOfMessageNotificationSoundsSelect,
};

export const mobileNotificationInputFieldData = (defaultOption: string, formatMessage: IntlShape['formatMessage']): FieldsetRadio => {
    const messages = defineMessages({
        allMessages: {
            id: 'channelNotifications.mobileNotification.newMessages',
            defaultMessage: 'All new messages {optionalDefault}',

        },
        mention: {
            id: 'channelNotifications.mobileNotification.mention',
            defaultMessage: 'Mentions, direct messages, and keywords only {optionalDefault}',
        },
        nothing: {
            id: 'channelNotifications.mobileNotification.nothing',
            defaultMessage: 'Nothing {optionalDefault}',
        },
    });

    return {
        options: [
            {
                dataTestId: `MobileNotification-${NotificationLevels.ALL}`,
                title: (
                    <FormattedMessage
                        {...messages.allMessages}
                        values={{
                            optionalDefault: defaultOption === NotificationLevels.ALL ? (
                                <FormattedMessage
                                    {...defaultMessage}
                                />) : undefined,
                        }}
                    />
                ),

                name: formatMessage(messages.allMessages, {
                    optionalDefault: defaultOption === NotificationLevels.ALL ? (
                        formatMessage(defaultMessage)
                    ) : undefined,
                }),
                key: `MobileNotification-${NotificationLevels.ALL}`,
                value: NotificationLevels.ALL,
            },
            {
                dataTestId: `MobileNotification-${NotificationLevels.MENTION}`,
                title: (
                    <FormattedMessage

                        {...messages.mention}
                        values={{
                            optionalDefault: defaultOption === NotificationLevels.MENTION ? (
                                <FormattedMessage
                                    {...defaultMessage}
                                />) : undefined,
                        }}
                    />
                ),
                name: formatMessage(messages.mention, {
                    optionalDefault: defaultOption === NotificationLevels.MENTION ? (
                        formatMessage(defaultMessage)
                    ) : undefined,
                }),
                key: `MobileNotification-${NotificationLevels.MENTION}`,
                value: NotificationLevels.MENTION,
            },
            {
                dataTestId: `MobileNotification-${NotificationLevels.NONE}`,
                title: (
                    <FormattedMessage
                        {...messages.nothing}
                        values={{
                            optionalDefault: defaultOption === NotificationLevels.NONE ? (
                                <FormattedMessage
                                    {...defaultMessage}
                                />) : undefined,
                        }}
                    />
                ),
                name: formatMessage(messages.nothing, {
                    optionalDefault: defaultOption === NotificationLevels.NONE ? (
                        formatMessage(defaultMessage)
                    ) : undefined,
                }),
                key: `MobileNotification-${NotificationLevels.NONE}`,
                value: NotificationLevels.NONE,
            },
        ],
    };
};

export default {
    desktopNotificationInputFieldData,
    desktopNotificationSoundsCheckboxFieldData,
    desktopNotificationSoundsSelectFieldData,
    IgnoreMentionsInputFieldData,
    mobileNotificationInputFieldData,
    MuteChannelInputFieldData,
    DesktopReplyThreadsInputFieldData,
    MobileReplyThreadsInputFieldData,
    AutoFollowThreadsInputFieldData,
    sameMobileSettingsDesktopInputFieldData,
};
