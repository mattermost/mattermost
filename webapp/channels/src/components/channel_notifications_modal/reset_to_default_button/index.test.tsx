// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import type {ChannelMembership} from '@mattermost/types/channels';
import type {UserNotifyProps} from '@mattermost/types/users';

import {renderWithContext, screen} from 'tests/react_testing_utils';
import {NotificationLevels, DesktopSound} from 'utils/constants';
import {notificationSoundKeys, convertDesktopSoundNotifyPropFromUserToDesktop} from 'utils/notification_sounds';
import {TestHelper} from 'utils/test_helper';

import ResetToDefaultButton, {SectionName} from './index';
import type {Props} from './index';

describe('ResetToDefaultButton', () => {
    const defaultProps: Props = {
        sectionName: SectionName.Desktop,
        userNotifyProps: TestHelper.getUserMock().notify_props,
        userSelectedChannelNotifyProps: TestHelper.getChannelMembershipMock({}).notify_props,
        onClick: jest.fn(),
    };

    test('should not show if section name is not valid', () => {
        // Mock console.error to suppress PropTypes warning
        const originalError = console.error;
        console.error = jest.fn();

        const props = {
            ...defaultProps,
            sectionName: 'invalidSectionName' as any,
        };

        const {container} = renderWithContext(<ResetToDefaultButton {...props}/>);
        expect(container).toBeEmptyDOMElement();

        console.error = originalError;
    });

    test('should render if section name is valid', () => {
        const {container} = renderWithContext(<ResetToDefaultButton {...defaultProps}/>);
        expect(container).not.toBeEmptyDOMElement();
        expect(container).toMatchSnapshot();
    });

    test('should not show if desktop notifications are same as default user notification settings', () => {
        const props = {
            ...defaultProps,
            userNotifyProps: {
                ...defaultProps.userNotifyProps,
                desktop: NotificationLevels.ALL,
                desktop_threads: NotificationLevels.ALL,
                desktop_sound: 'true' as UserNotifyProps['desktop_sound'],
                desktop_notification_sound: notificationSoundKeys[0] as UserNotifyProps['desktop_notification_sound'],
            },
            userSelectedChannelNotifyProps: {
                ...defaultProps.userSelectedChannelNotifyProps,
                desktop: NotificationLevels.ALL,
                desktop_threads: NotificationLevels.ALL,
                desktop_sound: DesktopSound.ON,
                desktop_notification_sound: notificationSoundKeys[0] as ChannelMembership['notify_props']['desktop_notification_sound'],
            },
        };

        const {container} = renderWithContext(<ResetToDefaultButton {...props}/>);
        expect(container).toBeEmptyDOMElement();
    });

    test('should show if desktop notifications are not same as user notification settings', () => {
        const props1 = {
            ...defaultProps,
            userNotifyProps: {
                ...defaultProps.userNotifyProps,
                desktop: NotificationLevels.ALL,
                desktop_threads: NotificationLevels.ALL,
                desktop_sound: 'true' as UserNotifyProps['desktop_sound'],
                desktop_notification_sound: notificationSoundKeys[0] as UserNotifyProps['desktop_notification_sound'],
            },
            userSelectedChannelNotifyProps: {
                ...defaultProps.userSelectedChannelNotifyProps,
                desktop: NotificationLevels.MENTION,
                desktop_threads: NotificationLevels.ALL,
                desktop_sound: DesktopSound.ON,
            },
        };

        const {rerender} = renderWithContext(<ResetToDefaultButton {...props1}/>);
        expect(screen.getByText('Reset to default')).toBeInTheDocument();

        const props2 = {
            ...props1,
            userSelectedChannelNotifyProps: {
                ...defaultProps.userSelectedChannelNotifyProps,
                desktop: NotificationLevels.ALL,
                desktop_sound: DesktopSound.OFF,
                desktop_notification_sound: notificationSoundKeys[0] as UserNotifyProps['desktop_notification_sound'],
            },
        };

        rerender(<ResetToDefaultButton {...props2}/>);
        expect(screen.getByText('Reset to default')).toBeInTheDocument();
    });

    test('should not show if mobile notifications are same as default user notification settings', () => {
        const props = {
            ...defaultProps,
            sectionName: SectionName.Mobile,
            userNotifyProps: {
                ...defaultProps.userNotifyProps,
                push: NotificationLevels.ALL,
                push_threads: NotificationLevels.ALL,
            },
            userSelectedChannelNotifyProps: {
                ...defaultProps.userSelectedChannelNotifyProps,
                push: NotificationLevels.ALL,
                push_threads: NotificationLevels.ALL,
            },
        };

        const {container} = renderWithContext(<ResetToDefaultButton {...props}/>);
        expect(container).toBeEmptyDOMElement();
    });

    test('should show if mobile notifications are not same as user notification settings', () => {
        const props = {
            ...defaultProps,
            sectionName: SectionName.Mobile,
            userNotifyProps: {
                ...defaultProps.userNotifyProps,
                push: NotificationLevels.ALL,
                push_threads: NotificationLevels.ALL,
            },
            userSelectedChannelNotifyProps: {
                ...defaultProps.userSelectedChannelNotifyProps,
                push: NotificationLevels.MENTION,
                push_threads: NotificationLevels.ALL,
            },
        };

        const {rerender} = renderWithContext(<ResetToDefaultButton {...props}/>);
        expect(screen.getByText('Reset to default')).toBeInTheDocument();

        const props2 = {
            ...defaultProps,
            userSelectedChannelNotifyProps: {
                ...defaultProps.userSelectedChannelNotifyProps,
                push: NotificationLevels.ALL,
                push_threads: NotificationLevels.MENTION,
            },
        };

        rerender(<ResetToDefaultButton {...props2}/>);
        expect(screen.getByText('Reset to default')).toBeInTheDocument();
    });
});

describe('convertDesktopSoundNotifyPropFromUserToDesktop', () => {
    test('should return ON if user desktop sound is true', () => {
        const desktopSound = convertDesktopSoundNotifyPropFromUserToDesktop('true');
        expect(desktopSound).toBe(DesktopSound.ON);
    });

    test('should return ON if user desktop sound is not present', () => {
        const desktopSound = convertDesktopSoundNotifyPropFromUserToDesktop();
        expect(desktopSound).toBe(DesktopSound.ON);
    });

    test('should return OFF if user desktop sound is false', () => {
        const desktopSound = convertDesktopSoundNotifyPropFromUserToDesktop('false');
        expect(desktopSound).toBe(DesktopSound.OFF);
    });
});
