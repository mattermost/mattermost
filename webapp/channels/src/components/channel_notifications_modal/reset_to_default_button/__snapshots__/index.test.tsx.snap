// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetToDefaultButton should render if section name is valid 1`] = `
<div>
  <button
    class="channel-notifications-settings-modal__reset-btn"
    data-testid="resetToDefaultButton-desktop"
  >
    <i
      class="icon icon-refresh"
    />
    Reset to default
  </button>
</div>
`;
